app/models/workers_scaling/algorithm_factory.rb

```diff
@@ -28,19 +28,57 @@ def self.get_algorithms
     end
 
     ##
-    # Arguments:
-    # * name - symbol representing Algorithm implementation
-    # * experiment,
-    #   user_id,
-    #   allowed_infrastructures,
-    #   planned_finish_time - standard attributes for Algorithms (see #SampleAlgorithm for example)
-    # * params - hash with additional parameters for specific Algorithm implementations
+    # Arguments: attributes hash containing:
+    #  * name - symbol representing Algorithm implementation
+    #  * experiment_id,
+    #    user_id,
+    #    allowed_infrastructures,
+    #    planned_finish_time,
+    #    last_update_time - standard attributes for Algorithms (see #Algorithm for details)
+    #  * params (optional) - hash with additional parameters for specific Algorithm implementations
     # Returns new instance of Algorithm implementation for given name
     # Raises AlgorithmNameUnknown if given name is unknown
-    # Raises AlgorithmParameterMissing if attributes required by Algorithm implementation are not present in params
-    def self.create_algorithm(name, experiment, user_id, allowed_infrastructures, planned_finish_time, params = {})
-      raise AlgorithmNameUnknown unless ALGORITHMS.keys.include? name
-      ALGORITHMS[name].new(experiment, user_id, allowed_infrastructures, planned_finish_time, params)
+    # Raises AlgorithmParameterMissing if attributes required by Algorithm are not present
+    def self.create_algorithm(attributes)
+      attributes.symbolize_keys!
+      validate_attributes(attributes)
+      class_name = attributes[:name].to_sym
+      raise AlgorithmNameUnknown.new unless ALGORITHMS.keys.include? class_name
+      ALGORITHMS[class_name].new(attributes).initialize_runtime_fields
+    end
+
+
+    ##
+    # Returns instance of algorithm from cache
+    # If Algorithm is not stored in cache, new one will be created
+    # If stored algorithm has last_update_time older than the one in database,
+    # it will be deleted from cache
+    def self.get_algorithm(experiment_id)
+      cache_key = "workers_scaling_algorithm_#{experiment_id}"
+      if Rails.cache.exist?(cache_key)
+        last_update_time = Algorithm.where({experiment_id: experiment_id},
+                                           fields: [:last_update_time])
+                               .first.last_update_time
+        if Rails.cache.read(cache_key).last_update_time < last_update_time
```

Comment:
Jeśli masz aktualny wpis do wyciągasz go 2 razy zawsze. Nie jest to może jakoś bardzo obciążające ale dałoby się to raz wyciągać.